op-node/rollup/addresses/addresses.go

```
@@ -0,0 +1,56 @@
+package addresses
+
+import (
+	"math/big"
+
+	"github.com/ethereum-optimism/optimism/op-service/predeploys"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+type CeloAddresses struct {
+	SuggestedFeeRecipient common.Address
+}
+
+var (
+	MainnetAddresses = &CeloAddresses{
+		SuggestedFeeRecipient: common.HexToAddress("0xaa"), // TODO: replace it with the actual one
+	}
+
+	AlfajoresAddresses = &CeloAddresses{
+		SuggestedFeeRecipient: predeploys.SequencerFeeVaultAddr,
+	}
+
+	BaklavaAddresses = &CeloAddresses{
+		SuggestedFeeRecipient: predeploys.SequencerFeeVaultAddr,
```

I would prefer any other address than Optimism's default here. If our change doesn't apply, I want to notice on each testnet!

---

I think to change the recipient address for the testnets, we need to do it during the next hardfork (maybe Holocene?) since the coinbase address affects the block hash.

---

I still don't see what would break if we changed it outside of a hardfork. If I have some time left, I can change it on my local alfajores node and see what happens.

---

I changed these and successfully synced a baklava node with the resulting op-node. So this can be changed without hardforking.

---

I'm still curious about how the `miner`(coinbase) field in the block header does not affect the block hash. Do you know how it is excluded when calculating the block hash?

Also I'm curious when we change `SuggestedFeeRecipient` address for Alfajores or Baklava and run the node, if there were any non-deposit transactions, the tip will go to the new address instead of `SequencerFeeVault`, which changes the state data in terms of balance, isn't it? (make difference from sequencer state data)

---

Changing the FeeRecipient absolutely changes the hash. But I think the verifying nodes will use the FeeRecipient set by the block builder when verifying the block, so they both arrive at the same result. I'll see if I find that part of the code.

---

The beneficiary (coinbase) gets set in https://github.com/celo-org/op-geth/blob/9e9144c99e64445376481930093b56c6fb8630fd/core/evm.go#L53 which in our case gets [set to the coinbase from the header](https://github.com/celo-org/op-geth/blob/9e9144c99e64445376481930093b56c6fb8630fd/consensus/beacon/consensus.go#L81). The header's coinbase is not further validated anywhere.

Any specific piece of code you think would fail? I'm never really sure about these details myself. It isn't easy to find all the right places in the code.

---

I thought the change of coinbase address in the verifier node might end up with discrepancy with the sequencer because the batch data only includes L2 transaction data, not block header data like coinbase address. In batch derivation code, the verifier node decodes batch data and prepare attributes data by itself and attach only the L2 transaction data in batch to the prepared attributes data. ([here](https://github.com/celo-org/optimism/blob/celo-rebase-12/op-node/rollup/derive/attributes_queue.go#L124-L132) and [here](https://github.com/celo-org/optimism/blob/celo-rebase-12/op-node/rollup/derive/attributes.go#L168-L177))

And then if the verifier node sets different address for suggested fee recipient, then geth will construct the block header with it, resulting the different block hash. ([code](https://github.com/celo-org/op-geth/blob/9e9144c99e64445376481930093b56c6fb8630fd/core/types/block.go#L74-L131))

But you verified successful syncing when changing coinbase address, so I have no idea what I am missing..

---

I agree that the [line setting `SuggestedFeeRecipient` in `PreparePayloadAttributes`](https://github.com/celo-org/optimism/blob/b3f0d13a580746946c7d58d63d33eba0e4720705/op-node/rollup/derive/attributes.go#L171) looks like something should break. I'll try to find out why my experiment still worked.

---

I think it worked because the verified payload was not constructed locally but received via
```go
// From op-node/rollup/driver/state.go                                                                                                                                                                                                                           
func (s *Driver) OnUnsafeL2Payload(ctx context.Context, envelope *eth.ExecutionPayloadEnvelope) error {                                                                                                                                                          
    payload := envelope.ExecutionPayload // Comes from network                                                                                                                                                                                                   
    // Validation happens here                                                                                                                                                                                                                                   
}
 ```
 Once the `FeeRecipient` is set in the payload, it is not checked or overridden.
 ```go
// op-service/eth/types.go:                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            
func (envelope *ExecutionPayloadEnvelope) CheckBlockHash() (common.Hash, bool) {                                                                                                                                                                                 
    // Uses THE PAYLOAD'S FeeRecipient to compute hash                                                                                                                                                                                                           
    header := types.Header{                                                                                                                                                                                                                                      
        Coinbase: payload.FeeRecipient, // From received block                                                                                                                                                                                                   
        // ...                                                                                                                                                                                                                                                   
    }                                                                                                                                                                                                                                                            
    return header.Hash() == payload.BlockHash // Validates internal consistency                                                                                                                                                                                  
}  
 ```
 
Is `PreparePayloadAttributes` ever used if the node is not the sequencer?

---

Oh yes this process is for unsafe sync between sequencer and other verifier nodes for the unsafe blocks (not too old blocks, not submitted as batch yet). In unsafe sync, you are right. The verifier node receives the envelope built by sequencer and just use the fee recipient address.

Otherwise for the old blocks, already submitted as batch, the verifier nodes should derive the batch data from L1, not from unsafe sync. In that scenario, `PreparePayloadAttributes` is used to construct the payload and set the fee recipient address by itself.

---

Yes, agreed. This difference probably masked the actual failure:
* Syncing unsafe blocks works.
* Geth does did not log any informative errors about this, instead `missing trie node` errors appeared which I wrongly attributed to bad shutdowns or other data corruption.
* Errors about safe blocks appeared significantly after being fully synced in my test (which was even harder to notice due to my doing a clean resync after the trie node error).
* But op-node eventually logs `op-node-1  | t=2025-03-11T15:02:07+0000 lvl=error msg="Critical error" err="failed to process block with only deposit transactions: payload attributes are not valid, cannot build block: Invalid payload attributes"`

So you we're right all along and this really is a hardfork (on OP-Stack, I still think it would not be on normal Ethereum but that's irrelevant for us). Thanks for being persistent on this! I'll take a look at the code again to check that this error message really is caused by the address change and I'm not fooled by something else again.

---

Thanks for verifying it! The only way to avoid hardfork is to use the predeploy contract, I guess. If we want to change the address by hardcoding it, or adding a flag for it, regardless of the method, a hardfork would inevitably be required.

---

Yes. We could also still undo this change as long is its not running on any network and rely on the ProxyAdmin to change the SequencerFeeVault implementation at a later time. I didn't really consider that contract change option earlier because I wanted to avoid changing the contracts on short notice, but the option of using the SequencerFeeVault's address and postponing the contract change sounds quite appealing now.

---

Yep agreed indeed! We need to request primitives team to check the feasibility of upgrading the contract to handle other ERC20s and then we can just revert this changes.